### PR TITLE
Fix S3 gateway cross-repo copies

### LINKED ISSUE #7467: Support cross-repository copy and MPU copy in S3 gateway

### PR DESCRIPTION
**Pull request by arielshaqed**
putobject and friends were confusing src and dst repositories, which will not do.

Fixes #7467.

### CODE REVIEW

**Comment by arielshaqed** [comment]
**Reviewers:** Adding relevant test to Esti, but please review the change for now.

**Comment by arielshaqed** [comment]
> **Reviewers:** Adding relevant test to Esti, but please review the change for now.

Actually... testing this may not be possible with the current S3 gateway test, which uses the MinIO client, which I am not sure supports the multipart copies API.  May take a while, will open a tech-debt issue if I cannot get this done here today.

**Comment by arielshaqed** [comment]
> @arielshaqed Thanks, code is great no comments on the changes. However, can you please add a test scenario for that

Thanks!  There are [some issues with testing](https://github.com/treeverse/lakeFS/pull/7468#issuecomment-1946232289), I'm still trying to figure out if I can use MinIO client ComposeObject to do this.

**Comment by N-o-Z** [comment]
> may not be pos



> > **Reviewers:** Adding relevant test to Esti, but please review the change for now.
> 
> Actually... testing this may not be possible with the current S3 gateway test, which uses the MinIO client, which I am not sure supports the multipart copies API. May take a while, will open a tech-debt issue if I cannot get this done here today.

@arielshaqed for that specific test you can use an S3 client and limit the test for s3 blockstore only, though I think minio does support multipart upload

**Comment by arielshaqed** [comment]
@guy-har maybe you know?  I managed to get Esti to pass, except for...
1. SigV2.  I am still missing some headers to sign.  This is less urgent, I doubt we have many users of MPU copy with sigV2.  So I am happy to open a tech debt issue and pull without it.
1. Azure tests fail due to "InvalidBlockList".  This would need to be an error in the Azure block adapter.  The obvious suspect is the conversion from symmetrical bounds in the S3 API (bytes [start,end]) to asymmetrical bounds in the Azure BlobStore API (bytes [offset, offset+count)).  But I had a look and we do add 1 to get the required bounds.  Any ideas what else to try?

**Review by arielshaqed — COMMENTED** [review]
@N-o-Z I added an integration test to Esti.  That was a great call because it uncovered an issue with the Azure block adapter!  So I fixed that one too.  Please re-re-review.

@guy-har I'd love to have your review of everything, especially of the changes to that Azure block adapter :-)

Please note that sigv2 still doesn't work.  Opened #7472 to do that.  But we should pull anyway: nobody should use sigV2, it's really hard to do nowadays, it solves the original problem for Boto3/Python users who are currently complaining, and lakeFS with this code is better than without it!

File: esti/s3_gateway_test.go [hunk 1]
```diff
@@ -328,6 +328,89 @@ func TestS3HeadBucket(t *testing.T) {
 	})
 }
 
+func TestS3CopyObjectMultipart(t *testing.T) {
+	const minDataContentLengthForMultipart = 5 << 20
```
Comment (guy-har) [inline esti/s3_gateway_test.go:332]: nit: it's a bit confusing that `minDataContentLengthForMultipart` and `largeDataContentLength` exist in different files and only used here

**Review by guy-har — APPROVED** [review]
Looks Great!
Thanks!

File: pkg/testutil/random.go [hunk 2]
```diff
@@ -33,3 +35,27 @@ func RandomString(rand *rand.Rand, size int) string {
...
+		n = int(r.remaining)
+	}
+	// n still fits into int!
+	r.rand.Read(p[:n])
+	r.remaining -= int64(n)
+	return n, nil
+}
+
+// RandomReader returns a reader that will return size bytes from rand.
+func RandomReader(rand *rand.Rand, size int64) io.Reader {
```
Comment (N-o-Z) [inline pkg/testutil/random.go:59]:
```suggestion
func NewRandomReader(rand *rand.Rand, size int64) io.Reader {
```
?

**Review by N-o-Z — APPROVED** [review]
Thanks for adding the test!!
Some optional suggestion

File: pkg/block/azure/multipart_block_writer.go [hunk 3]
```diff
@@ -187,25 +198,20 @@ func copyPartRange(ctx context.Context, destinationContainer container.Client, d
...
 	if err != nil {
 		return nil, fmt.Errorf("failed staging part data: %w", err)
 	}
 
 	// stage size data
-	sizeData := fmt.Sprintf("%d\n", size)
-	blobSizesURL := destinationContainer.NewBlockBlobClient(destinationObjName + sizeSuffix)
-	_, err = blobSizesURL.StageBlock(ctx, base64Etag, streaming.NopCloser(strings.NewReader(sizeData)), nil)
+	sizeData := fmt.Sprintf("%d\n", count)
+	blobSizesBlob := destinationContainer.NewBlockBlobClient(destinationKey.BlobURL + sizeSuffix)
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/block/azure/multipart_block_writer.go:213]: Not done in this PR.

File: esti/s3_gateway_test.go [hunk 4]
```diff
@@ -328,6 +328,88 @@ func TestS3HeadBucket(t *testing.T) {
 	})
 }
 
+func TestS3CopyObjectMultipart(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply esti/s3_gateway_test.go:331]: The only thing tests are good for are finding bugs.  And fixing them.  And preventing regressions.  Automatically.  Not sure it's worth the effort.

Although... in this case this test actually found bugs.  And helped fix one of them.  And another test in the file prevented a regression.  Automatically.  Actually the case against writing tests has never been weaker 🥳 .

File: pkg/testutil/random.go (hunk 2 above)
Comment (arielshaqed) [reply pkg/testutil/random.go]: Nice one, yes please!

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Pulling once all tests pass.

**Comment by arielshaqed** [comment]
license/cla stuck.  But **I work here**, I _literally signed **on paper** to grant the project my rights for work on lakeFS_.  Pulling by adminiforce.

**Pull request merged**